syntax = "proto3";

package strobes.analyzer;

import "google/protobuf/struct.proto";

message ScanMeta {
  int32 critical = 1;
  int32 high = 2;
  int32 medium = 3;
  int32 low = 4;
  int32 info = 5;
  int32 total_bugs = 6;
}

message SAST {
  string file_name = 1;
  string vulnerable_code = 2;
  string start_line_no = 3;
  string end_line_no = 4;    
}

message DAST {
  repeated string uri = 1;
  int32 port = 2;
  string method = 3;
  string param = 4;
  string evidence = 5;
}

message Network {
  string ipaddress = 1;
  string port = 2;
  string hostname = 3;
  string macaddress = 4;
  string cpe = 5;
}

message Package {
  repeated string affected_versions = 1;
}

message Container {
  string package_name = 1;
  string fixed_version = 2;
  string installed_version = 3;
  string cpe = 4;
}

message AWS {
  enum Type {
    AWS_ACCOUNT = 0;
    AWS_API_GATEWAY_REST_API = 1;
    AWS_CERTIFICATE_MANAGER_CERTIFICATE =2;
    AWS_CLOUD_FORMATION_STACK = 3;
    AWS_CLOUD_FRONT_DISTRIBUTION = 4;
    AWS_CLOUDTRAIL_TRAIL = 5;
    AWS_EC2_EIP = 6;
    AWS_EC2_INSTANCE = 7;
    AWS_EC2_SECURITY_GROUP = 8;
    AWS_EC2_SNAPSHOT = 9;
    AWS_EC2_VOLUME = 10;
    AWS_EC2_VPC = 11;
    AWS_ECS_TASK_DEFINITION = 12;
    AWS_EKS_CLUSTER = 13;
    AWS_ELASTICSEARCH_DOMAIN = 14;
    AWS_ELB_LOADBALANCER = 15;
    AWS_IAM_POLICY = 16;
    AWS_IAM_ROLE = 17;
    AWS_IAM_USER = 18;
    AWS_KMS_KEY = 19;
    AWS_LAMBDA_FUNCTION = 20;
    AWS_RDSDB_INSTANCE = 21;
    AWS_RDSDB_SNAPSHOT = 22;
    AWS_REDSHIFT_CLUSTER = 23;
    AWS_S3_BUCKET = 24;
    AWS_SNS_TOPIC = 25;
    AWS_SQS_QUEUE = 26;
    OTHERS = 27;
  }
  string arn = 1;
  string partition = 2;
  string resource_id = 3;
  Type type = 4;
  string category = 5;
  repeated string resources = 6;
  string vulnerable_id = 7;
}

message Cloud {
  string region = 1;
  AWS aws = 2;
}

message Bug {
  string title = 1;
  string description = 2;
  int32 severity = 3;
  float CVSS = 4;
  string step_to_reproduce = 5;
  string mitigation = 6;
  repeated string CVE = 7;
  repeated string CWE = 8;
  repeated string references = 9;
  SAST sast = 10;
  DAST dast = 11;
  Network network = 12;
  Cloud cloud = 13;
  Package package = 14;
  Container container = 15;
  map<string, string> scan_details = 16;
  map<string, string> assets = 17;
  repeated string tags = 18;
  google.protobuf.Struct extra_info = 19;
}

message Result {
    ScanMeta scan_meta = 1;
    repeated Bug bugs_list = 2;
    enum Analyzer {
      SPOTBUGS = 0;
      PROWLER = 1;
      ZAP = 2;
      TRIVY = 3;
      ANCHORE = 4;
      BANDIT = 5;
      BRAKEMAN = 6;
      NODEJSCAN = 7;
      PHPCS = 8;
      ESLINT = 9;
      SCODESCAN = 10;
      GITLEAKS = 11;
      CLOUDMAPPER = 12;
      GOSEC = 13;
      FLAWFINDER = 14;
    }
    Analyzer analyzer = 4;
}
